single_kernel_mongo/status.py

```diff
@@ -39,17 +68,93 @@ def to_waiting(self, message: str):
-        if message is None:
-            self.set_and_share_status(ActiveStatus())
-            return
-        self.set_and_share_status(ActiveStatus(message))
-
     def to_maintenance(self, message: str):
         """Sets status to maintenance."""
         self.set_and_share_status(MaintenanceStatus(message))
 
     def to_error(self, message: str):
         """Sets status to error."""
         self.set_and_share_status(ErrorStatus(message))
+
+    def get_statuses(self) -> StatusesDict:
+        """Collects the statuses of all managers."""
+        if self.operator.name == CharmKind.MONGOD:
+            return StatusesDict(
+                {
+                    "mongodb": self.operator.mongo_manager.get_status(),
+                    "shard": self.operator.shard_manager.get_status(),
+                    "config-server": self.operator.config_server_manager.get_status(),
+                    "PBM": self.operator.backup_manager.get_status(),
+                }
+            )
+        return StatusesDict(
+            {
+                "mongodb": WaitingStatus("waiting for mongos to start")
+                if not self.operator.workload.active
+                else ActiveStatus(),
+                "shard": None,
+                "config-server": None,
+                "PBM": None,
+            }
+        )
+
+    def prioritize_statuses(self, statuses: StatusesDict) -> StatusBase:
+        """Prioritizes the statuses."""
+        mongodb_status, shard_status, config_server_status, pbm_status = (
+            statuses["mongodb"],
+            statuses["shard"],
+            statuses["config-server"],
+            statuses["PBM"],
+        )
+        if not isinstance(mongodb_status, ActiveStatus):
+            return mongodb_status
+
+        if shard_status and not isinstance(shard_status, ActiveStatus):
+            return shard_status
+
+        if config_server_status and not isinstance(config_server_status, ActiveStatus):
+            return config_server_status
+
+        if pbm_status and not isinstance(pbm_status, ActiveStatus):
+            return pbm_status
+
+        # if all statuses are active report mongodb status over sharding status
+        # This is also always the case for mongos charms because the 3 final statuses are None.
+        return mongodb_status
+
+    def process_and_share_statuses(self) -> None:
+        """Retrieves statuses from processes inside charm and shares the highest priority status.
+
+        When a non-fatal error occurs while processing statuses, the error is processed and
+        returned as a statuses.
+        """
+        # retrieve statuses of different services running on Charmed MongoDB
+        deployment_mode = (
+            "replica set" if self.state.is_role(MongoDBRoles.REPLICATION) else "cluster"
+        )
+        waiting_status = None
+        try:
+            statuses = self.get_statuses()
+        except OperationFailure as e:
+            if e.code in [13, 18]:
```

Comment:
[can we keep the original constants](https://github.com/canonical/mongodb-operator/blob/6306245b18b2627d521f50a7595c36035bcd47fb/lib/charms/mongodb/v0/set_status.py#L182-L184)